libs/server-sdk/src/data_store/dependency_tracker.hpp

```diff
@@ -0,0 +1,165 @@
+
+namespace launchdarkly::server_side::data_store {
+
+/**
+ * Class which can be used to tag a collection with the DataKind that collection
+ * is for. This is primarily to decrease the complexity of iterating collections
+ * allowing for a kvp style iteration, but with an array storage container.
+ * @tparam Storage
+ */
+template <typename Storage>
+class TaggedData {
+   public:
+    explicit TaggedData(DataKind kind) : kind_(kind) {}
+    [[nodiscard]] DataKind Kind() const { return kind_; }
+    [[nodiscard]] Storage& Data() { return storage_; }
+
+   private:
+    DataKind kind_;
+    Storage storage_;
+};
+
+/**
+ * Class used to maintain a set of dependencies. Each dependency may be either
+ * a flag or segment.
+ * For instance, if we have a flagA, which has a prerequisite of flagB, and
+ * a segmentMatch targeting segmentA, then its dependency set would be
+ * ```
+ * [{DataKind::kFlag, "flagB"}, {DataKind::kSegment, "segmentA"}]
+ * ```
+ */
+class DependencySet {
+   public:
+    DependencySet();
+    using DataType = std::array<TaggedData<std::set<std::string>>,
+                                static_cast<std::size_t>(DataKind::kKindCount)>;
+    void Set(DataKind kind, std::string key);
+
+    void Remove(DataKind kind, std::string const& key);
+
+    [[nodiscard]] bool Contains(DataKind kind, std::string const& key);
+
+    [[nodiscard]] std::set<std::string> SetForKind(DataKind kind);
+
+    /**
+     * Return the size of all the data kind sets.
+     * @return The combined size of all the data kind sets.
+     */
+    [[nodiscard]] std::size_t Size();
+
+    [[nodiscard]] typename DataType::iterator begin();
+
+    [[nodiscard]] typename DataType::iterator end();
+
+   private:
+    DataType data_;
+};
+
+/**
+ * Class used to map flag/segments to their set of dependencies.
+ * For instance, if we have a flagA, which has a prerequisite of flagB, and
+ * a segmentMatch targeting segmentA, then a dependency map, containing
+ * this set, would be:
+ * ```
+ * {{DataKind::kFlag, "flagA"}, [{DataKind::kFlag, "flagB"},
+ *  {DataKind::kSegment, "segmentA"}]}
+ * ```
+ */
+class DependencyMap {
+   public:
+    DependencyMap();
+    using DataType =
+        std::array<TaggedData<std::unordered_map<std::string, DependencySet>>,
+                   static_cast<std::size_t>(DataKind::kKindCount)>;
+    void Set(DataKind kind, std::string key, DependencySet val);
+
+    [[nodiscard]] std::optional<DependencySet> Get(DataKind kind,
+                                                   std::string const& key);
+
+    void Clear();
+
+    [[nodiscard]] typename DataType::iterator begin();
+
+    [[nodiscard]] typename DataType::iterator end();
+
+   private:
+    DataType data_;
+};
+
+/**
+ * This class implements a mechanism of tracking dependencies of flags and
+ * segments. Both the forward dependencies (flag A depends on flag B) but also
+ * the reverse (flag B is depended on by flagA).
+ */
+class DependencyTracker {
```

Comment:
This dependency tracker works basically the same as the one in node and dotnet.

Reply:
There are lots of strings in this representation. It is probably fine, but something to consider if we have any performance issues.